Expand InputType to support number type

darknight opened the pull request:
This PR is for issue https://github.com/slint-ui/slint/issues/3356

Changes include:

- Update enum `InputType`, add `Number` and `Decimal` field
- Update `key_event` logic
- Add two more test cases

ogoffart reviewed (approved):
Thanks a lot for the patch. Looks good.

(It's just missing a ChangeLog entry)

darknight commented:
> Thanks a lot for the patch. Looks good.
> 
> (It's just missing a ChangeLog entry)

Thanks for the review.

Updated `CHANGELOG.md`, added one more log entry for this change under `Changelog -> Unreleased -> Slint Language` section.

darknight commented on internal/core/items/text.rs:
```diff
@@ -479,6 +479,11 @@ impl Item for TextInput {
                 {
                     return KeyEventResult::EventIgnored;
                 }
+                if matches!(self.input_type(), InputType::Number)
```
I put this check after `text.insert_str(insert_pos, &event.text);` because I need the updated `text` to verify if it is valid decimal.

For example, user inputs `1, 2, 3`, then move cursor to the head, input `-`, then move cursor to the head again & input `-`, this can be detected and disallowed because the parsing for `--123` will fail.

If the check is placed after the shortcuts but before the `text.insert_str(...)`, although I have the info of `self.text()` and `event.text` but I don't have the info of insert position, what I can only do is assume `event.text` will append to `self.text()`.

darknight replied:
@ogoffart Need your input here for my concern :)

ogoffart replied:
I think it should go after the shortcut handling, together with the decimal handling.

ogoffart commented on CHANGELOG.md:
```diff
@@ -28,6 +28,7 @@ All notable changes to this project are documented in this file.
  - Added `item-pointer-event` to `StandardListView`.
  - Added `orientation` property to `Slider`.
  - Added experimental `cupertino` style.
+ - Added `Number`, `Decimal` variant to enum `InputType`
```
(the changelog entry must be moved to the next release)
